pkg/target/http.go

```diff
@@ -402,26 +413,28 @@ func (ht *HTTPTarget) Write(messages []*models.Message) (*models.TargetWriteResu
 
 			response := response{Body: string(responseBody), Status: resp.StatusCode}
 
-			if response.Status >= 200 && response.Status < 300 {
+			if findMatchingRule(response, ht.responseRules.Invalid) != nil {
 				for _, msg := range goodMsgs {
-					if msg.AckFunc != nil { // Ack successful messages
-						msg.AckFunc()
-					}
-					sent = append(sent, msg)
+					// can we use response body as an error message for invalid data?
+					msg.SetError(errors.New(response.Body))
 				}
-				continue
-			}
 
-			if findMatchingRule(response, ht.responseRules.Invalid) != nil {
 				invalid = append(invalid, goodMsgs...)
 				continue
 			}
 
-			if findMatchingRule(response, ht.responseRules.SetupError) != nil {
+			var errorDetails error
+			if rule := findMatchingRule(response, ht.responseRules.SetupError); rule != nil {
```

Comment:
No objection to this implementation, but a note for reference that using [a switch](https://golangdocs.com/switch-statement-in-golang) is an option for this kind of logic too

You can have conditions depend on several things, whihc can often result in readability improvements

Reply:
Not sure if in this particular case switch would improve readability :/ But if you have some neat idea how to improve this part, I'd love to see it 😁